modules/runner/src/main/java/org/apache/ignite/internal/app/IgniteServerImpl.java

```diff
@@ -82,10 +86,33 @@ public class IgniteServerImpl implements IgniteServer {
 
     private final ClassLoader classLoader;
 
-    private volatile @Nullable IgniteImpl instance;
+    private final Executor asyncContinuationExecutor;
+
+    /** Current Ignite instance. This field is not volatile to make hot path accesses from IgniteReference and other references
+     * faster (they always happen under a read lock, which guarantees visibility of changes to this field). So we access
+     * this field in this object under synchronization ({@link #igniteMonitor} serves as the monitor).
+     */
+    private @Nullable IgniteImpl ignite;
```

Comment:
I noticed that we only read and write in this field, why not replace it with `volatile`?

Reply:
I tried to explain this in the javadoc. When executing user operations, we always read this under a read lock, so we don't need it to be `volatile` for visibility. If we make it `volatile`, we will add an extra volatile read, which might hurt the performance.

Reply:
If I understood the documentation correctly, it says that reading from the synchronization block is cheaper in terms of performance than one volatile read.
Is there any proof of this? If we have two parallel threads reading this field, they will step on each other's toes and park someone, which will lead to performance degradation.

Reply:
I tried making the field volatile and rerun the benchmarks. Volatility effect is not noticable at all. So I just simplified everything by making the field volatile and removing `syncrhonized`.